linera-core/src/client/mod.rs

```diff
@@ -1979,11 +1977,18 @@ where
                 blobs,
             ))
         };
-        // Check the final block proposal. This will be cheaper after #1401.
-        self.client
-            .local_node
-            .handle_block_proposal(*proposal.clone())
-            .await?;
+        if manager
+            .requested_proposed
+            .as_ref()
+            .map(|proposal| &proposal.content)
+            != Some(&proposal.content)
```

Comment:
I fail to see how this addition is related to the main objective of the PR - can you explain?

Reply:
Ok, but we extracted the `requested_propsed` earlier already - in line 1950 - can we use that?

Reply:
> We now already call handle_block_proposal earlier, before we set the pending block to that proposal.

Earlier than what? I can't find that it changed from what it used to be.

Reply:
I don't see how we could merge the two places where we look into `requested_proposed`; the logic is very different, and in the second case, the condition is also true in the `None` case.

> Earlier than what?

I mean we call it already before we set the pending block, in `execute_block`.

Reply:
> I mean we call it already before we set the pending block, in execute_block.

Now I see it - in `execute_block` we call 
- `set_pending_block` which calls 
- - `handle_block_proposal` and then 
- - `self.state_mut().set_pending_block(block)`.
-  _After_ that we call `process_pending_block_without_prepare` which in turn calls 
- - this`propose_block`. 
 
So we don't need to re-execute.

It's all so clear now 🙃